eclipse/src/saros/filesystem/EclipseReferencePointImpl.java

```diff
@@ -0,0 +1,280 @@
+package saros.filesystem;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import org.eclipse.core.runtime.CoreException;
+
+/** Eclipse implementation of the Saros reference point interface. */
+// TODO rename to EclipseReferencePoint
+public class EclipseReferencePointImpl implements IReferencePoint {
+
+  /** The delegate represented by the reference point. */
+  private final org.eclipse.core.resources.IContainer delegate;
+
+  /**
+   * Instantiates an Eclipse reference point object.
+   *
+   * @param delegate the delegate resource represented by the reference point
+   * @throws NullPointerException if the given delegate is <code>null</code>
+   * @throws IllegalArgumentException if the given delegate does not exist
+   */
+  public EclipseReferencePointImpl(org.eclipse.core.resources.IContainer delegate) {
+    Objects.requireNonNull(delegate, "The given delegate must not be null");
+
+    if (!delegate.exists()) {
+      throw new IllegalArgumentException("The given delegate must exist - " + delegate);
+    }
+
+    this.delegate = delegate;
+  }
+
+  /**
+   * Returns the delegate represented by this reference point.
+   *
+   * @return the delegate represented by this reference point
+   */
+  org.eclipse.core.resources.IContainer getDelegate() {
+    return delegate;
+  }
+
+  @Override
+  public EclipsePathImpl getReferencePointRelativePath() {
+    return new EclipsePathImpl(org.eclipse.core.runtime.Path.EMPTY);
+  }
+
+  @Override
+  public boolean exists() {
+    return delegate.exists();
+  }
+
+  @Override
+  public String getName() {
+    return delegate.getName();
+  }
+
+  @Override
+  public boolean exists(IPath path) {
+    Objects.requireNonNull(path, "Given path must not be null");
+
+    return delegate.exists(((EclipsePathImpl) path).getDelegate());
+  }
+
+  @Override
+  public List<IResource> members() throws IOException {
```

Comment:
I decided against having an abstract implementation of `IContainer` that both the folder and reference point implementation extend. My first prototype implementation did it this way, but it made the logic much harder to understand/change. As the two implementations now operate on different instance variables, implementing shared parent classes gets very cumbersome as the logic can't access any of the instance variables directly.

(Furthermore, the logic is not _exactly_ the same. It differs in the handling of the relative child paths of the members. The reference point implementation just uses the names as the patch while the folder implementation appends the names to their reference-point-relative path. This could theoretically be unified by also appending to the reference-point-relative path for reference points as it is always empty, but this makes the logic unnecessarily harder to understand.)

I thought about splitting `IReferencePoint` from the `IResource`/`IContainer` interface to make this split more apparent. In general, there aren't many places where we want to be able to interchangeably operate on container representing shared folders or reference points. Reference points are only needed to determine shared content. Most (if not all) other cases should just be interested in shared folders. So, from a design-standpoint, it would be sensible to split the implementations to make this split in purpose/usage more apparent. But, as this is only an assumption on my part and I didn't want to look through all usages of `IFolder` and `IReferencePoint` to check whether a split would be possible and further increase the scope, I do not intend to split them as part of this migration.